crates/ruff_python_parser/src/lexer.rs

```diff
@@ -508,6 +511,138 @@ impl<'source> Lexer<'source> {
+                        }
+                    }
+                    // Consume the escaped character.
+                    self.cursor.bump();
+                }
+                quote @ ('\'' | '"') if quote == context.quote_char() => {
+                    match context.quote_size {
+                        StringQuoteSize::Single => break,
+                        StringQuoteSize::Triple => {
+                            let mut remaining = self.cursor.rest().chars().skip(1);
+                            if remaining.next() == Some(quote) && remaining.next() == Some(quote) {
+                                break;
+                            }
+                        }
+                    }
+                    self.cursor.bump();
+                }
+                '{' => {
+                    if self.cursor.second() == '{' {
+                        self.cursor.bump();
+                        self.cursor.bump();
+                    } else {
+                        break;
+                    }
+                }
+                '}' => {
+                    if in_named_unicode {
+                        in_named_unicode = false;
+                        self.cursor.bump();
+                    } else if self.cursor.second() == '}' && !context.is_in_format_spec() {
+                        self.cursor.bump();
+                        self.cursor.bump();
+                    } else {
+                        break;
+                    }
+                }
+                _ => {
+                    self.cursor.bump();
+                }
+            }
+        }
+        let value_end = self.offset();
+
+        if value_start == value_end {
+            return Ok(None);
+        }
+
+        let value = self.source[TextRange::new(value_start, value_end)]
+            .to_string()
+            .replace("{{", "{")
+            .replace("}}", "}");
+        Ok(Some(Tok::FStringMiddle(value)))
+    }
+
+    fn maybe_lex_fstring_end(&mut self) -> Option<Tok> {
+        // SAFETY: Safe because the function is only called when `self.fstring_stack` is not empty.
+        let context = self.fstring_stack.last().unwrap();
+
+        if self.cursor.first() == context.quote_char() {
+            match context.quote_size {
+                StringQuoteSize::Single => {
+                    self.cursor.bump();
+                    return Some(Tok::FStringEnd);
+                }
+                StringQuoteSize::Triple => {
+                    let mut remaining = self.cursor.rest().chars().skip(1);
+                    if remaining.next() == Some(context.quote_char())
+                        && remaining.next() == Some(context.quote_char())
+                    {
+                        self.cursor.bump();
+                        self.cursor.bump();
+                        self.cursor.bump();
+                        return Some(Tok::FStringEnd);
+                    }
+                }
+            }
+        }
```

Comment:
I prefer to use `eat` whenever possible because it avoids bugs where I accidentally forgot to call `bump`, which results in an infinite loop:

```suggestion
        match context.quote_size {
            StringQuoteSize::Single => {
                if self.cursor.eat_char(context.quote_char()) {
                    return Some(Tok::FStringEnd);
                }
            }
            StringQuoteSize::Triple => {
                let expected = match context.quote_char {
                    StringQuoteChar::Single => "'''",
                    StringQuoteChar::Double => r#"""""#,
                };
                if self.cursor.rest().starts_with(expected) {
                    self.cursor.bump();
                    self.cursor.bump();
                    self.cursor.bump();
                    return Some(Tok::FStringEnd);
                }
            }
        }
```

I like @konstin's suggestion to add a `eat_char2` and `eat_char3` that only eat if all two or three chars match.

Reply:
Oh yes, I can totally relate to the part where we forgot to call `bump` :)

I've added these methods.